pretyping/typing.ml

```diff
@@ -403,13 +403,20 @@ let judge_of_array env sigma u tj defj tyj =
   in
   sigma, j
 
-let bad_relevance_msg = CWarnings.create_msg Typeops.bad_relevance_warning ()
+type ('constr,'types,'r) bad_relevance =
+| BadRelevanceBinder of 'r * ('constr,'types,'r) Context.Rel.Declaration.pt
+| BadRelevanceCase of 'r * 'constr
+
+let bad_relevance_warning =
+  CWarnings.create_warning ~name:"bad-relevance" ~default:CWarnings.AsError ()
+
+let bad_relevance_msg = CWarnings.create_msg bad_relevance_warning ()
 (* no need for default printer, pretyping is always linked with himsg in practice *)
 
 let warn_bad_relevance_binder ?loc env sigma rlv bnd =
-  match CWarnings.warning_status Typeops.bad_relevance_warning with
+  match CWarnings.warning_status bad_relevance_warning with
 | CWarnings.Disabled | CWarnings.Enabled ->
-  CWarnings.warn bad_relevance_msg ?loc (env,sigma,Typeops.BadRelevanceBinder (rlv, bnd))
+  CWarnings.warn bad_relevance_msg ?loc (env,sigma,BadRelevanceBinder (rlv, bnd))
 | CWarnings.AsError ->
   CErrors.anomaly (CErrors.print (PretypeError (env, sigma, TypingError (Type_errors.BadBinderRelevance (rlv, bnd)))))
```

Comment:
I am really confused by this function. Why is it an anomaly and not just an error? Does that mean that the user cannot trigger it (but a plugin could)? Also, what is the rationale for allowing the user to disable an anomaly?

Reply:
I agree this should be an error. The convention in the kernel is not consistent, sometimes we throw an error at ill-formed terms, sometimes we fail. I think that given that this code can be called by the checker, an anomaly should really happen only when there is a programming error, not when the kernel gets some ill-formed input.

Reply:
I have no memory of making this an anomaly but blame says it was me.

@ppedrot note that this is pretyping/typing not the kernel.

Reply:
(My remark also goes for the upper layers, especially since we have Ltac2 to poke around terms and break invariants.)

Reply:
https://github.com/coq/coq/pull/19195